### ISSUE TITLE
Possible thread safety issue in BTree destructor

### ISSUE THREAD

**Issue by langston-barrett** [souffle-lang/souffle#1687]
I compiled some of the Souffle tests with Thread Sanitizer and got many reports. Here's one of them:
```
==================
WARNING: ThreadSanitizer: data race (pid=17160)
  Read of size 8 at 0x7b0c00000098 by main thread:
    #0 std::unique_ptr<souffle::t_btree_i__0__1, std::default_delete<souffle::t_btree_i__0__1> >::~unique_ptr() /home/langston/code/souffle/src/../include/souffle/datastructure/BTree.h:1935 (souffletbxTYD+0x425a80)
    #1 souffle::Sf_souffletbxTYD::~Sf_souffletbxTYD() souffletbxTYD.cpp:261 (souffletbxTYD+0x425a80)
    #2 main souffletbxTYD.cpp:521 (souffletbxTYD+0x4088c5)

  Previous write of size 8 at 0x7b0c00000098 by thread T7:
    #0 souffle::detail::btree<souffle::Tuple<int, 1ul>, souffle::t_btree_i__0__1::t_comparator_0, std::allocator<souffle::Tuple<int, 1ul> >, 256u, souffle::detail::binary_search, true, souffle::t_btree_i__0__1::t_comparator_0, souffle::detail::updater<souffle::Tuple<int, 1ul> > >::insert(souffle::Tuple<int, 1ul> const&, souffle::detail::btree<souffle::Tuple<int, 1ul>, souffle::t_btree_i__0__1::t_comparator_0, std::allocator<souffle::Tuple<int, 1ul> >, 256u, souffle::detail::binary_search, true, souffle::t_btree_i__0__1::t_comparator_0, souffle::detail::updater<souffle::Tuple<int, 1ul> > >::btree_operation_hints<1u>&) /home/langston/code/souffle/src/../include/souffle/datastructure/BTree.h:1344 (souffletbxTYD+0x448d17)
    #1 souffle::t_btree_i__0__1::insert(souffle::Tuple<int, 1ul> const&, souffle::t_btree_i__0__1::context&) souffletbxTYD.cpp:137 (souffletbxTYD+0x44a669)
    #2 souffle::Sf_souffletbxTYD::subroutine_3(std::vector<int, std::allocator<int> > const&, std::vector<int, std::allocator<int> >&)::{lambda()#1}::operator()() const [clone ._omp_fn.0] souffletbxTYD.cpp:474 (souffletbxTYD+0x44a669)
    #3 <null> <null> (libgomp.so.1+0x1a805)

  Location is heap block of size 40 at 0x7b0c00000090 allocated by main thread:
    #0 operator new(unsigned long) <null> (libtsan.so.0+0x761aa)
    #1 std::unique_ptr<souffle::t_btree_i__0__1, std::default_delete<souffle::t_btree_i__0__1> > souffle::mk<souffle::t_btree_i__0__1, souffle::t_btree_i__0__1>() /home/langston/code/souffle/src/../include/souffle/utility/ContainerUtil.h:43 (souffletbxTYD+0x41896a)
    #2 souffle::Sf_souffletbxTYD::Sf_souffletbxTYD() souffletbxTYD.cpp:255 (souffletbxTYD+0x42ae27)
    #3 main souffletbxTYD.cpp:521 (souffletbxTYD+0x4087a9)

  Thread T7 (tid=17168, running) created by main thread at:
    #0 pthread_create <null> (libtsan.so.0+0x3055b)
    #1 <null> <null> (libgomp.so.1+0x1ae25)
    #2 souffle::Sf_souffletbxTYD::runFunction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool) souffletbxTYD.cpp:294 (souffletbxTYD+0x44b501)
    #3 souffle::Sf_souffletbxTYD::runAll(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) souffletbxTYD.cpp:303 (souffletbxTYD+0x44b751)
    #4 main souffletbxTYD.cpp:526 (souffletbxTYD+0x40887c)

SUMMARY: ThreadSanitizer: data race /home/langston/code/souffle/src/../include/souffle/datastructure/BTree.h:1935 in std::unique_ptr<souffle::t_btree_i__0__1, std::default_delete<souffle::t_btree_i__0__1> >::~unique_ptr()
```
The report points to this method:

https://github.com/souffle-lang/souffle/blob/81fc43a24a208f4d1f33ffc41b31f327a70ef645/src/include/souffle/datastructure/BTree.h#L1931-L1944

This does indeed seem fishy, because there is a lock that seems to guard modifications to the root node, but isn't acquired by the descructor/`clear`:

https://github.com/souffle-lang/souffle/blob/81fc43a24a208f4d1f33ffc41b31f327a70ef645/src/include/souffle/datastructure/BTree.h#L1221-L1225

Is this an issue, or am I misunderstanding the purpose of that lock and/or it's interactions with the destructor?

**Comment by b-scholz** [comment]
We have parallel code in the generated C++ object. However, the whole C++ on its own is not thread-safe, i.e., we assume it is invoked by a single thread. 

The BTree destructor is one example - it will happen when the generated object is destructed at the end. 

It looks to me that you have multiple threads that execute the generated C++ object - is this correct?

**Comment by langston-barrett** [comment]
@b-scholz I'm not sure I understand your comments (please bear with me - I'm a novice in multithreaded C++).

> The BTree destructor is one example - it will happen when the generated C++ object is destructed at the end.

Does this mean that there should be one instance of `BTree` per invocation of the generated code, and it will be in the scope of the "main"/calling thread? So the idea is that while most operations on a `BTree` must be synchronized, the constructor and destructor need not be?

> It looks to me that you have multiple threads that execute the generated C++ object - is this correct?

Not as far as I know: The above trace was collected from running some of the `FastExecution` unit tests, so the generated code is called by Souffle (with the `-c` flag).

**Comment by b-scholz** [comment]
We generate a C++ class that does the computation of the logic program. The C++ object runs some portions of the code in parallel. Other portions are sequential. 

The issue now is that clear() and the destruction of Btree should be in the sequential portion of the code and not in the parallel one. 

Hence, I am quite puzzled that you see these messages.  My explanation would have been that you use parallel external functors or use the C++ interface. 

However, you are testing stand-alone. Can you give us the details of your test-case so that we can reproduce the error messages?

**Comment by mmcgr** [comment]
The OpenMP library can confuse the thread sanitiser since they both attempt to modify the same code. That could be the issue here.

**Comment by langston-barrett** [comment]
I'm on vacation at the moment, I'll post a reproducible diff when I'm back!

**Comment by langston-barrett** [comment]
This patch and Dockerfile will exhibit a bunch of Thread Sanitizer errors in the test suite:
```dockerfile
FROM ubuntu:20.04
RUN apt-get update -y -qq && apt-get install -y -qq git autoconf automake bison build-essential clang doxygen flex g++ git libncurses5-dev libtool libsqlite3-dev make mcpp python sqlite zlib1g-dev
COPY ./patch /tmp/patch
RUN git clone --depth=1 --branch=master https://github.com/souffle-lang/souffle && \
    cd souffle && \
    git apply /tmp/patch && \
    ./bootstrap && \
    ./configure --enable-debug && \
    make -j24
WORKDIR souffle/tests
ENV SOUFFLE_CATEGORY=Evaluation
CMD make check-local -j24 SOUFFLE_CATEGORY=Evaluation TESTSUITEFLAGS=--errexit || bash -c "cat testsuite.dir/**/{*.out,*.err}"
```
```diff
diff --git a/src/souffle-compile.in b/src/souffle-compile.in
index d5477b52b..83e5d17ae 100644
--- a/src/souffle-compile.in
+++ b/src/souffle-compile.in
@@ -46,8 +46,8 @@ error() {
 CXX="$(printenv CXX || true)"
 test -z "$CXX" && CXX="@CXX@"
 CPPFLAGS="$(printenv CPPFLAGS || true) -std=c++17 @CPPFLAGS@"
-CXXFLAGS="$(printenv CXXFLAGS || true) @SOUFFLE_CXXFLAGS@ -march=native"
-LDFLAGS="$(printenv LDFLAGS || true) @LDFLAGS@"
+CXXFLAGS="$(printenv CXXFLAGS || true) @SOUFFLE_CXXFLAGS@ -O1 -march=native -fsanitize=thread"
+LDFLAGS="$(printenv LDFLAGS || true) @LDFLAGS@ -fsanitize=thread"
 LIBS="$(printenv LIBS || true) @LIBS@"
 
 # set by command flags
@@ -105,6 +105,7 @@ error "cannot open source file: '$1'" $?
 exe=`basename $1 .cpp`
 test "$1" != "$exe"
 error "source file is not a .cpp file: '$1'" $?
+cp $1 /tmp/souffle-compile.cpp
 
 # Ensure binary is compiled to same directory as cpp file
 cd "$(dirname $1)"
```
Just copy those to your current directory and run
```
docker build -t souffle-tsan .
docker run --rm -it souffle-tsan
```
I found [some tips](https://stackoverflow.com/questions/33004809/can-i-use-thread-sanitizer-for-openmp-programs#35619738) online for reducing false positives with TSan + OpenMP, I'll try integrating those next to see if these errors disappear.

**Comment by langston-barrett** [comment]
The following Dockerfile incorporates those suggestions from StackOverflow, but still raises a lot of error messages:
```dockerfile
FROM ubuntu:20.04                              
ENV DEBIAN_FRONTEND=noninteractive
RUN apt-get update -y -qq && apt-get install -y -qq git autoconf automake bison build-essential clang doxygen flex git libncurses5-dev libtool libsqlite3-dev make mcpp python sqlite zlib1g-dev cmake clang
COPY ./patch /tmp/patch
ENV CC=clang
ENV CXX=clang++
RUN git clone --depth=1 --branch=master https://github.com/llvm/llvm-project && \
    cd llvm-project && \
    mkdir build && \
    cd build && \
    cmake -DLIBOMP_TSAN_SUPPORT=1 ../openmp && \
    cmake --build . -j24 --target install
RUN git clone --depth=1 --branch=master https://github.com/souffle-lang/souffle && \
    cd souffle && \
    git apply /tmp/patch && \
    ./bootstrap && \
    ./configure --enable-debug && \
    make -j24
WORKDIR souffle/tests
ENV SOUFFLE_CATEGORY=Evaluation
RUN ls /usr/local/lib/libomp.so
ENV LD_PRELOAD=/usr/local/lib/libomp.so
CMD make check-local -j24 SOUFFLE_CATEGORY=Evaluation TESTSUITEFLAGS=--errexit || bash -c "cat testsuite.dir/**/{*.out,*.err}"
```
Another answer on that SO thread suggested using a tool called Archer, which has evidently been merged into mainline LLVM as of LLVM 10: https://github.com/PRUNERS/archer#deprecated. Might be worth a shot to try that as well.

**Comment by HerbertJordan** [comment]
I managed to reproduce the issue and investigated some of the reports.

These data races are correctly reported. They are indeed present -- and intended behavior.

The optimistic locking mechanism of souffle's BTrees implementation is based on something called a [Seqlock](https://en.wikipedia.org/wiki/Seqlock). Instead of realizing a conflict avoidance strategy, this mechanism relies on conflict detection and resolution (similar to some transactional memory or database systems).

In particular: Instead of guaranteeing shared read-only or exclusive write access during data manipulations -- which would cause high lock contention and pure performance in souffle's BTree case -- the Seqlock allows 'racy' access and facilitates a mechanism to validate that performed memory reads/writes performed during e.g. an insert operation are not conflicting with concurrently processed operations.

C++ provides ways to express such semantic safely. However, it appears to be in conflict with the thread sanitizer, most likely assuming synchronization through controlling exclusive access to memory locations.

Thus, I would categorize this as intended behavior, no fix required. A potential mitigation step would be to exempt BTree manipulations and other optimistic locking based operations from being covered by the thread sanitizer. See [here](https://clang.llvm.org/docs/SanitizerSpecialCaseList.html) for details.

**Comment by langston-barrett** [comment]
@HerbertJordan Okay, thanks for looking into it!
